src/source.rs

```diff
@@ -186,7 +173,119 @@ impl std::fmt::Debug for FileSource {
             .field("name", &self.name)
             .field("langids", &self.langids)
             .field("pre_path", &self.pre_path)
-            .field("cache", &self.cache)
             .finish()
     }
 }
+
+impl Inner {
+    fn lookup_resource<F>(&self, path: String, f: F) -> ResourceStatus
+    where
+        F: FnOnce() -> ResourceStatus,
+    {
+        let mut lock = self.entries.borrow_mut();
+        lock.entry(path).or_insert_with(|| f()).clone()
+    }
+
+    fn update_resource(&self, path: String, resource: RcResourceOption) -> RcResourceOption {
+        let mut lock = self.entries.borrow_mut();
+        let entry = lock.get_mut(&path);
+        match entry {
+            Some(entry) => *entry = resource.clone().into(),
+            _ => panic!("Expected "),
+        }
+        resource
+    }
+
+    pub fn has_file(&self, full_path: &str) -> Option<bool> {
```

Comment:
If I'm not mistaken the `None` here means `we don't know if the file exists - it's either currently being loaded, or we never tried`, can you add a doc string about it?

Reply:
Done